src/common.ts

```diff
@@ -54,4 +57,56 @@ export class Common {
     logger.debug(`loginUrl: ${loginUrl}`);
     return loginUrl;
   }
+
+  public static async identifyPossibleScratchOrgs(fields: AuthFields, orgAuthInfo: AuthInfo): Promise<void> {
+    const logger = await Logger.child('Common', { tag: 'identifyPossibleScratchOrgs' });
+
+    // return if we already know the hub or we know it is a devhub or prod-like
+    if (fields.isDevHub || fields.devHubUsername) return;
+    // there are no hubs to ask, so quit early
+    if (!(await AuthInfo.hasAuthentications())) return;
+    logger.debug('getting devHubs from authfiles');
+
+    // TODO: return if url is not sandbox-like to avoid constantly asking about production orgs
+    // TODO: someday we make this easier by asking the org if it is a scratch org
+
+    const hubAuthInfos = await this.getDevHubAuthInfos();
+    logger.debug(`found ${hubAuthInfos.length} DevHubs`);
+    if (hubAuthInfos.length === 0) return;
+
+    // ask all those orgs if they know this orgId
+    await Promise.all(
+      hubAuthInfos.map(async (hubAuthInfo) => {
+        try {
+          const devHubOrg = await Org.create({ aliasOrUsername: hubAuthInfo.getUsername() });
+          const conn = devHubOrg.getConnection();
+          const data = await conn.query<QueryResult<{ Id: string }>>(
+            `select Id from ScratchOrgInfo where ScratchOrg = '${sfdc.trimTo15(orgAuthInfo.getFields().orgId)}'`
+          );
+          if (data.totalSize > 0) {
+            // if any return a result
+            logger.debug(`found orgId ${fields.orgId} in devhub ${hubAuthInfo.getUsername()}`);
+            try {
+              await orgAuthInfo.save({ ...orgAuthInfo.getFields(), devHubUsername: hubAuthInfo.getUsername() });
+            } catch (error) {
+              logger.debug(`error updating auth file for ${orgAuthInfo.getUsername()}`, error);
+            }
+            logger.debug(`set ${hubAuthInfo.getUsername()} as devhub for scratch org ${orgAuthInfo.getUsername()}`);
+          }
+        } catch (error) {
+          logger.error(`Error connecting to devhub ${hubAuthInfo.getUsername()}`, error);
+        }
+      })
+    );
+  }
+
+  public static async getDevHubAuthInfos(): Promise<AuthInfo[]> {
+    return (
+      await Promise.all(
+        (await AuthInfo.listAllAuthFiles())
+          .map((fileName) => basename(fileName, '.json'))
+          .map((username) => AuthInfo.create({ username }))
```

Comment:
I couldn't come up with any other way to figure out what auths are hubs.

Reply:
this approach limits the # of orgs that get queried in the above `Promise.all`

Reply:
the fastest alternative would be to read the auth files directly (not go through core) but that opens up its own set of problems.

We could also add a method in core to deliver the raw auth file contents as json instead of AuthInfos.  It's still potentially a lot of fs, but we could parallelize the read/parse better than core does today

Reply:
one more idea (kinda hacky) is to filter out obvious scratch-org-type names (foo@example.com, that thing where a timestamp is added to the main user to create secondary users, etc).